pkg/querier/block_streaming_test.go

```diff
@@ -0,0 +1,244 @@
+					values: []testSample{{6, 2}, {9, 2}},
+				},
+				{
+					lbls:   labels.FromStrings("foo", "bar3"),
+					values: []testSample{{3, 3}},
+				},
+			},
+			expResult: []testSeries{
+				{
+					lbls:   labels.FromStrings("foo", "bar1"),
+					values: []testSample{{1, 1}, {2, 1}, {5, 10}, {6, 2}, {9, 2}},
+				},
+				{
+					lbls:   labels.FromStrings("foo", "bar3"),
+					values: []testSample{{3, 3}},
+				},
+			},
+		},
+		"multiple entries of the same series again": {
+			input: []testSeries{
+				{
+					lbls:   labels.FromStrings("foo", "bar1"),
+					values: []testSample{{1, 1}, {2, 1}, {5, 10}},
+				},
+				{
+					lbls:   labels.FromStrings("foo", "bar1"),
+					values: []testSample{{6, 2}, {9, 2}},
+				},
+				{
+					lbls:   labels.FromStrings("foo", "bar3"),
+					values: []testSample{{3, 3}},
+				},
+				{
+					lbls:   labels.FromStrings("foo", "bar3"),
+					values: []testSample{{4, 3}, {5, 3}, {6, 3}},
+				},
+			},
+			expResult: []testSeries{
+				{
+					lbls:   labels.FromStrings("foo", "bar1"),
+					values: []testSample{{1, 1}, {2, 1}, {5, 10}, {6, 2}, {9, 2}},
+				},
+				{
+					lbls:   labels.FromStrings("foo", "bar3"),
+					values: []testSample{{3, 3}, {4, 3}, {5, 3}, {6, 3}},
+				},
+			},
+		},
+		"multiple unique series but with erroring chunk streamer": {
+			errorChunkStreamer: true,
+			input: []testSeries{
+				{
+					lbls:   labels.FromStrings("foo", "bar1"),
+					values: []testSample{{1, 1}, {2, 1}, {5, 10}},
+				},
+				{
+					lbls:   labels.FromStrings("foo", "bar2"),
+					values: []testSample{{2, 2}, {9, 2}},
+				},
+				{
+					lbls:   labels.FromStrings("foo", "bar3"),
+					values: []testSample{{3, 3}},
+				},
+			},
+			expResult: []testSeries{
+				{
+					lbls: labels.FromStrings("foo", "bar1"),
+				},
+				{
+					lbls: labels.FromStrings("foo", "bar2"),
+				},
+				{
+					lbls: labels.FromStrings("foo", "bar3"),
+				},
+			},
+		},
+	}
+
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			ss := &blockStreamingQuerierSeriesSet{streamReader: &mockChunkStreamer{series: c.input, causeError: c.errorChunkStreamer}}
+			for _, s := range c.input {
+				ss.series = append(ss.series, &storepb.StreamingSeries{
+					Labels: mimirpb.FromLabelsToLabelAdapters(s.lbls),
+				})
+			}
+			idx := 0
+			var it chunkenc.Iterator
+			for ss.Next() {
+				s := ss.At()
+				require.Equal(t, c.expResult[idx].lbls, s.Labels())
+				it = s.Iterator(it)
+				if c.errorChunkStreamer {
+					require.Error(t, it.Err())
+					idx++
+					// If chunk streamer errors out, we still go through every
+					// series but we don't get any samples. So we continue here
+					// and check all the series.
+					continue
+				}
+				var actSamples []testSample
+				for it.Next() != chunkenc.ValNone {
+					ts, val := it.At()
+					actSamples = append(actSamples, testSample{t: ts, v: val})
+				}
+				require.Equal(t, c.expResult[idx].values, actSamples)
+				require.NoError(t, it.Err())
+				idx++
```

Comment:
[nit] Perhaps we could rejig this slightly to make the intention clearer?

```suggestion
				if c.errorChunkStreamer {
					require.Error(t, it.Err())
				} else {
					var actSamples []testSample
					for it.Next() != chunkenc.ValNone {
						ts, val := it.At()
						actSamples = append(actSamples, testSample{t: ts, v: val})
					}
					require.Equal(t, c.expResult[idx].values, actSamples)
					require.NoError(t, it.Err())
				}
				idx++
```